feat(models): Add ParentheticalGroup table and associated changes

ProbablyFaiz opened the pull request:
This will allow us down the line to include group representatives in search results etc.

We switch the views to now use lazily-computed ParentheticalGroup model objects. It is a 1:1 replacement for the current (crappy) caching regime but is much more flexible for the future.

mlissner commented:
I've just given it a skim so far, but:

 - SET_NULL surprised me. What's the goal there? I'll have to think more on that.

 - Did you learn something about the number of queries the summaries page was making (I saw all the prefetching)? If so, should we add a test to prevent regressions there? I find those kinds of regressions are very hard to avoid.

 - I want to find an automated semgrep rule about this (or similar), but I don't want to use db_index anymore (it's unofficially deprecated, but the Django folks haven't figured out how to get rid of it last I checked). Instead, can you please explicitly define any indexes using `Meta.indexes`?

ProbablyFaiz commented:
`SET_NULL` sets the `group_id` column of a parenthetical to null when the corresponding group is deleted, rather than deleting the parenthetical (as `CASCADE` would do). Often we will delete `ParentheticalGroup`s when we want to compute new groups for an opinion, but we don't want to wipe all the associated `Parenthetical`s away when we do that (we still need those!), we just want to remove their association with the group we're deleting.

ProbablyFaiz commented:
With respect to the prefetching, we were doing prefetching before, just in a method of the `OpinionCluster` model rather than here. I realized that we were accidentally loading in the full list of parentheticals even on the main opinion page because of that, so I moved the prefetch to the summaries view method so we only load all the extra data for that.

mlissner commented on cl/search/models.py:
```diff
@@ -2870,6 +2883,31 @@ class Meta:
+        on_delete=models.CASCADE,
+    )
+    score = models.FloatField(
+        db_index=True,
```
Sorry, if my comment about semgrep wasn't clear. Can you move db_index into the `Meta.indexes` attribute, please?

mlissner commented on cl/search/models.py:
```diff
@@ -2870,6 +2883,31 @@ class Meta:
 
+class ParentheticalGroup(models.Model):
+    opinion = models.ForeignKey(
+        Opinion, related_name="parenthetical_groups", on_delete=models.CASCADE
+    )
+    representative = models.ForeignKey(
+        Parenthetical,
+        related_name="represented_group",
+        on_delete=models.CASCADE,
+    )
```
Could you please add help_text to these fields too? It took me a minute to figure out what the structure was and help_text would have helped.

ProbablyFaiz commented:
Alright, this is ready for another look. Also removed the threshold adjustment and added group computation to the ingestion command so these groups don't need to be computed at page load anymore.

mlissner commented:
Thanks. This is fantastic. Merging!
